File: app/src/main/java/com/pietrantuono/podcasts/subscribedpodcasts/view/SubscribedPodcastFragment.java
package com.pietrantuono.podcasts.subscribedpodcasts.view;

import android.annotation.TargetApi;
import android.content.Intent;
import android.os.Build;
import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.v4.app.ActivityOptionsCompat;
import android.support.v4.app.Fragment;
import android.support.v4.app.FragmentManager;
import android.support.v4.util.Pair;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.LinearLayout;

import com.pietrantuono.podcasts.R;
import com.pietrantuono.podcasts.addpodcast.customviews.PodcastsRecycler;
import com.pietrantuono.podcasts.addpodcast.model.pojos.SinglePodcast;
import com.pietrantuono.podcasts.addpodcast.singlepodcast.view.SingleSubscribedPodcastActivity;
import com.pietrantuono.podcasts.main.view.MainActivity;
import com.pietrantuono.podcasts.main.view.TransitionsFramework;
import com.pietrantuono.podcasts.addpodcast.singlepodcast.view.AddSinglePodcastActivity;
import com.pietrantuono.podcasts.subscribedpodcasts.di.SubscribedPodcastModule;
import com.pietrantuono.podcasts.subscribedpodcasts.presenter.SubscribedPodcastPresenter;

import java.util.List;

import javax.inject.Inject;

import butterknife.BindView;
import butterknife.ButterKnife;

public class SubscribedPodcastFragment extends Fragment implements SubscribedPodcastView {
    private static final String TAG = (SubscribedPodcastFragment.class).getSimpleName();
    @Inject SubscribedPodcastPresenter presenter;
    @Inject TransitionsFramework transitionsFramework;
    @BindView(R.id.recycler) PodcastsRecycler recycler;

    public SubscribedPodcastFragment() {
    }

    public static void navigateTo(FragmentManager fragmentManager) {
        SubscribedPodcastFragment frag = (SubscribedPodcastFragment) fragmentManager.findFragmentByTag(SubscribedPodcastFragment.TAG);
        if (frag == null) {
            frag = new SubscribedPodcastFragment();
        }
        fragmentManager.beginTransaction().replace(R.id.fragmentContainer, frag, SubscribedPodcastFragment.TAG).commit();
    }

    @Override
    public void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        ((MainActivity) getActivity()).getMainComponent().with(new SubscribedPodcastModule()).inject(this);
    }

    @Nullable
    @Override
    public View onCreateView(LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
        View view = inflater.inflate(R.layout.fragment_subscribed, container, false);
        ButterKnife.bind(this, view);
        presenter.bindView(this);
        recycler.setListeners(presenter);
        return view;
    }

    @Override
    public void onError(Throwable throwable) {
    }

    @Override
    public void setPodcasts(List<SinglePodcast> list) {
        if (recycler.getAdapter().getItemCount() != list.size()) {
            recycler.setItems(list);
        }
    }

    @Override
    public void onResume() {
        super.onResume();
        presenter.onStart();
    }

    @Override
    public void onPause() {
        super.onPause();
        presenter.onStop();
    }

    @TargetApi(Build.VERSION_CODES.LOLLIPOP)
    @Override
    public void startDetailActivityWithTransition(SinglePodcast singlePodcast, ImageView
            imageView, LinearLayout titleContainer) {
        Intent intent = new Intent(getActivity(), SingleSubscribedPodcastActivity.class);
        intent.putExtra(AddSinglePodcastActivity.Companion.getSINGLE_PODCAST_TRACK_ID(), singlePodcast.getTrackId());
        intent.putExtra(AddSinglePodcastActivity.Companion.getSTARTED_WITH_TRANSITION(), true);
        getActivity().startActivity(intent, ActivityOptionsCompat.makeSceneTransitionAnimation(getActivity(), getPairs(imageView, titleContainer)).toBundle());
    }

    @TargetApi(Build.VERSION_CODES.LOLLIPOP)
    @Override
    public void startDetailActivityWithoutTransition(SinglePodcast singlePodcast) {
        Intent intent = new Intent(getActivity(), AddSinglePodcastActivity.class);
        intent.putExtra(AddSinglePodcastActivity.Companion.getSINGLE_PODCAST_TRACK_ID(), singlePodcast);
        getActivity().startActivity(intent);
    }

    @NonNull
    private Pair[] getPairs(ImageView imageView, LinearLayout titleContainer) {
        return transitionsFramework.getPairs(imageView, getActivity(), titleContainer);
    }

    @Override
    public boolean isPartiallyHidden(int position) {
        return recycler.isPartiallyHidden(position);
    }
}

File: app/src/main/java/com/pietrantuono/podcasts/subscribedpodcasts/view/SubscribedPodcastFragment.kt
package com.pietrantuono.podcasts.subscribedpodcasts.view

import android.annotation.TargetApi
import android.content.Intent
import android.os.Build
import android.os.Bundle
import android.support.v4.app.ActivityOptionsCompat
import android.support.v4.app.Fragment
import android.support.v4.app.FragmentManager
import android.support.v4.util.Pair
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.ImageView
import android.widget.LinearLayout
import butterknife.BindView
import butterknife.ButterKnife
import com.pietrantuono.podcasts.R
import com.pietrantuono.podcasts.addpodcast.customviews.PodcastsRecycler
import com.pietrantuono.podcasts.addpodcast.singlepodcast.view.EpisodesListActivity
import com.pietrantuono.podcasts.main.view.MainActivity
import com.pietrantuono.podcasts.main.view.TransitionsHelper
import com.pietrantuono.podcasts.subscribedpodcasts.di.SubscribedPodcastModule
import com.pietrantuono.podcasts.subscribedpodcasts.presenter.SubscribedPodcastPresenter
import com.pietrantuono.podcasts.utils.ARTWORK
import com.pietrantuono.podcasts.utils.SINGLE_PODCAST_TRACK_ID
import com.pietrantuono.podcasts.utils.STARTED_WITH_TRANSITION
import models.pojos.Podcast
import javax.inject.Inject

class SubscribedPodcastFragment : Fragment(), SubscribedPodcastView {
    @Inject lateinit var presenter: SubscribedPodcastPresenter
    @Inject lateinit var transitions: TransitionsHelper
    @BindView(R.id.recycler) lateinit var recycler: PodcastsRecycler

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        (activity as MainActivity).mainComponent?.with(SubscribedPodcastModule(activity))?.inject(this)
    }

    override fun onCreateView(inflater: LayoutInflater?, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        val view = inflater!!.inflate(R.layout.fragment_subscribed, container, false)
        ButterKnife.bind(this, view)
        return view
    }

    override fun onError(throwable: Throwable) {}

    override fun setPodcasts(list: List<Podcast>) {
        if (recycler.adapter.itemCount != list.size) {
            recycler.setItems(list)
        }
    }

    override fun onStart() {
        super.onStart()
        presenter.bindView(this)
        recycler.setListeners(presenter)
        presenter.onResume()
    }

    override fun onStop() {
        super.onStop()
        presenter.onPause()
        presenter.bindView(null)
    }

    @TargetApi(Build.VERSION_CODES.LOLLIPOP)
    override fun startDetailActivityWithTransition(podcast: Podcast, imageView: ImageView, titleContainer: LinearLayout) {
        val intent = getDetailActivityIntent(podcast)
        intent.putExtra(STARTED_WITH_TRANSITION, true)
        activity.startActivity(intent, ActivityOptionsCompat.makeSceneTransitionAnimation(activity, *getPairs(imageView, titleContainer)).toBundle())
    }

    override fun startDetailActivityWithoutTransition(podcast: Podcast) {
        activity.startActivity(getDetailActivityIntent(podcast))
    }

    private fun getDetailActivityIntent(podcast: Podcast): Intent {
        val intent = Intent(activity, EpisodesListActivity::class.java)
        intent.putExtra(SINGLE_PODCAST_TRACK_ID, podcast.trackId)
        intent.putExtra(ARTWORK, podcast.artworkUrl600)
        return intent
    }

    private fun getPairs(imageView: ImageView, titleContainer: LinearLayout): Array<Pair<View, String>?> {
        return transitions.getPairs(imageView, activity, titleContainer)
    }

    override fun isPartiallyHidden(position: Int): Boolean {
        return recycler.isPartiallyHidden(position)
    }

    companion object {
        private val TAG = SubscribedPodcastFragment::class.java.simpleName

        fun navigateTo(fragmentManager: FragmentManager) {
            var frag: SubscribedPodcastFragment? = fragmentManager.findFragmentByTag(SubscribedPodcastFragment.TAG) as? SubscribedPodcastFragment
            if (frag == null) {
                frag = SubscribedPodcastFragment()
            }
            fragmentManager.beginTransaction().replace(R.id.fragmentContainer, frag, SubscribedPodcastFragment.TAG).commit()
        }
    }
}
